src/ipld/util.rs

```diff
@@ -438,3 +447,233 @@ impl<DB: Blockstore, T: Iterator<Item = Tipset> + Unpin> Stream for ChainStream<
+///
+/// # Arguments
+///
+/// * `db` - A database that implements [`Blockstore`] interface.
+/// * `tipset_iter` - An iterator of [`Tipset`], descending order `$child -> $parent`.
+/// * `stateroot_limit` - An epoch that signifies how far back we need to inspect tipsets.
+/// in-depth. This has to be pre-calculated using this formula: `$cur_epoch - $depth`, where
+/// `$depth` is the number of `[`Tipset`]` that needs inspection.
+pub fn unordered_stream_chain<
+    DB: Blockstore + Sync + Send + 'static,
+    T: Iterator<Item = Tipset> + Unpin + Send + 'static,
+>(
+    db: Arc<DB>,
+    tipset_iter: T,
+    stateroot_limit: ChainEpoch,
+) -> UnorderedChainStream<DB, T> {
+    let (sender, receiver) = flume::bounded(2048);
+
+    let seen = Arc::new(Mutex::new(CidHashSet::default()));
+    let handle = UnorderedChainStream::start_workers(
+        db,
+        tipset_iter,
+        sender.clone(),
+        seen.clone(),
+        stateroot_limit,
+        false,
+    );
+
+    UnorderedChainStream {
+        seen: Arc::new(Mutex::new(CidHashSet::default())),
+        worker_handle: handle,
+        fail_on_dead_links: true,
+        block_receiver: receiver,
+        marker: Default::default(),
+    }
+}
+
+impl<
+        DB: Blockstore + Send + Sync + 'static,
+        T: Iterator<Item = Tipset> + Unpin + Send + 'static,
+    > UnorderedChainStream<DB, T>
+{
+    // TODO: sort out the DB to be able to pass it on to the workers.
+    fn start_workers(
+        db: Arc<DB>,
+        mut tipset_iter: T,
+        block_sender: Sender<anyhow::Result<Block>>,
+        seen: Arc<Mutex<CidHashSet>>,
+        stateroot_limit: ChainEpoch,
+        fail_on_dead_links: bool,
+    ) -> JoinHandle<anyhow::Result<()>> {
+        let handle = task::spawn(async move {
+            let mut handles = JoinSet::new();
+            let (extract_sender, extract_receiver) = flume::unbounded();
+            let (emit_sender, emit_receiver) = flume::unbounded();
+
+            for _ in 0..num_cpus::get() * 3 {
+                let seen = seen.clone();
+                let extract_receiver: flume::Receiver<Vec<Cid>> = extract_receiver.clone();
+                let extract_sender = extract_sender.clone();
+                let db = db.clone();
+                let block_sender = block_sender.clone();
+                handles.spawn(async move {
+                    while let Ok(mut cid_vec) = extract_receiver.recv_async().await {
+                        while let Some(cid) = cid_vec.pop() {
+                            if should_save_block_to_snapshot(cid) && seen.lock().insert(cid) {
+                                if let Some(data) = db.get(&cid)? {
+                                    if cid.codec() == fvm_ipld_encoding::DAG_CBOR {
+                                        let mut new_values = extract_cids(&data)?;
+                                        cid_vec.reserve(new_values.len());
```

Comment:
`Vec::append` already calls reserve.

Reply:
Yep, this use to push in a loop, needs to be removed. Thanks!